internal/rbd/migration.go

```diff
@@ -0,0 +1,118 @@
+	"github.com/ceph/ceph-csi/internal/util"
+	"github.com/ceph/ceph-csi/internal/util/log"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// isMigrationVolHandle validate the passed in volHash is a volume handle
+// of migrated volume.
+func isMigrationVolHandle(volHash string) bool {
+	return strings.Contains(volHash, migIdentifier) && strings.Contains(volHash, migImageNamePrefix)
+}
+
+// parseMigrationVolHandle decode the volume handle and generate a migrationHandle
+// struct which consists of version, mon, image name , pool and clusterID information.
+func parseMigrationVolHandle(vh string) (*migrationHandle, error) {
+	mh := &migrationHandle{}
+	handSlice := strings.Split(vh, migHandleFieldSep)
+	if len(handSlice) < migHandleTotalLength {
+		// its short of length in this case, so return error
+		return nil, ErrInvalidVolID
+	}
+	// Store pool
+	poolHash := strings.Join(handSlice[migHandleSplitLength:], migHandleFieldSep)
+	poolByte, dErr := hex.DecodeString(poolHash)
+	if dErr != nil {
+		return nil, ErrMissingPoolNameInVolID
+	}
+	mh.poolName = string(poolByte)
+	// Parse migration identifier, mons and image
+	for _, field := range handSlice[:migHandleSplitLength] {
+		switch {
+		case field == migIdentifier:
+			mh.identifier = true
+		case strings.Contains(field, migImageNamePrefix):
+			imageSli := strings.Split(field, migImageNamePrefix)
+			if len(imageSli) > 0 {
+				mh.imageName = migInTreeImagePrefix + imageSli[1]
+			}
+		case strings.Contains(field, migMonPrefix):
+			// ex: mons-7982de6a23b77bce50b1ba9f2e879cce
+			mh.clusterID = strings.Trim(field, migMonPrefix)
+		}
+	}
+	if mh.imageName == "" {
+		return nil, ErrMissingImageNameInVolID
+	}
+	if mh.poolName == "" {
+		return nil, ErrMissingPoolNameInVolID
+	}
+	if mh.clusterID == "" {
+		return nil, ErrDecodeClusterIDFromMonsInVolID
+	}
+
+	return mh, nil
+}
+
+// parseAndDeleteMigratedVolume get rbd volume details from the migration handle
+// and delete the volume from the cluster, return err if there was an error on the process.
+func parseAndDeleteMigratedVolume(ctx context.Context, volumeID string, cr *util.Credentials) error {
+	parsedMigHandle, err := parseMigrationVolHandle(volumeID)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to parse migration volume handle: %s , err: %v", volumeID, err)
+
+		return err
+	}
+	rv := &rbdVolume{}
+
+	// fill details to rv struct from parsed migration handle
+	rv.RbdImageName = parsedMigHandle.imageName
+	rv.Pool = parsedMigHandle.poolName
+	rv.ClusterID = parsedMigHandle.clusterID
+	rv.Monitors, err = util.Mons(util.CsiConfigFile, rv.ClusterID)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to fetch monitors using clusterID: %s", rv.ClusterID)
+	}
+
+	// connect to the volume.
+	err = rv.Connect(cr)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to get connected to the rbd image : %s", rv.RbdImageName)
+
+		return err
+	}
+	defer rv.Destroy()
+	// if connected , delete it
+	err = deleteImage(ctx, rv, cr)
```

Comment:
imageNotFound case is already handled somewhere? am not seeing it anywhere

Reply:
@Madhu-1 getImageID() returns ErrImageNotFound if provided image is not found. thats what I was referring :)

Reply:
if it returns `ErrImageNotFound` same will be returned from the `parseAndDeleteMigratedVolume` function and that will be treated as the internalserver error in `DeleteVolume`?

Reply:
@Madhu-1 yeah, thats what I intended with this version. In one way the delete request is broken it matches with `internal` server error expectation.  do you want to change it to specifically treat `ErrImageNotFound` and return `success`?

Reply:
Not sure why it's intended can you please elaborate more on this? What if the delete request is sent to csi driver multiple times? All subsequent requests will start failing and pv will never gets deleted? Is it against the csi specifications?

Reply:
yeah, good to treat that specific case as success.. thanks